py/desispec/pipeline/run.py

```diff
@@ -179,70 +232,128 @@ def run_task_list(tasktype, tasklist, opts, comm=None, db=None, force=False):
 
-        log.debug("Number of {} tasks ready to run is {} (total is {})".format(tasktype,len(runtasks),len(tasklist)))
+        log.info(
+            "Number of {} tasks ready to run is {} (total is {})".
+            format(tasktype, len(runtasks), len(tasklist))
+        )
 
     if comm is not None:
-        runtasks = comm.bcast(runtasks, root=0)
         ntask = comm.bcast(ntask, root=0)
         ndone = comm.bcast(ndone, root=0)
 
-    # Get the weights for each task.  Since this might trigger DB lookups, only
-    # the root process does this.
+    if ntask == 0:
+        # All tasks are done- return
+        return (ntask, ndone, 0)
 
-    weights = None
     if rank == 0:
-        weights = [ task_classes[tasktype].run_time(x, procs_per_node, db=db) \
-            for x in runtasks ]
-    if comm is not None:
-        weights = comm.bcast(weights, root=0)
-
-    # Now every process has the full list of tasks.  Get the max
-    # number of processes for this task type
+        # The root process queries the environment for DESI runtime variables
+        # set in pipeline-generated slurm scripts and uses default values if
+        # they are not found.  Then it computes the number of workers and the
+        # distribution of tasks in a way that is identical to what was done
+        # during job planning.
+
+        job_env = parse_job_env()
+        tfactor = 1.0
+        if "timefactor" in job_env:
+            tfactor = job_env["timefactor"]
+            log.info("Using timefactor {}".format(tfactor))
+        else:
+            log.warning(
+                "DESI_PIPE_RUN_TIMEFACTOR not found in environment, using 1.0."
+            )
+        startup = 0.0
+        if "startup" in job_env:
+            startup = job_env["startup"]
+            log.info("Using worker startup of {} minutes".format(startup))
+        else:
+            log.warning(
+                "DESI_PIPE_RUN_STARTUP not found in environment, using 0.0."
+            )
+        nworker = 0
+        if "workers" in job_env:
+            nworker = job_env["workers"]
+            log.info("Found {} workers from environment".format(nworker))
+        else:
+            taskproc = task_classes[tasktype].run_max_procs()
+            if taskproc == 0:
+                taskproc = procs_per_node
+            nworker = nproc // taskproc
+            if nworker == 0:
+                nworker = 1
+            log.warning(
+                "DESI_PIPE_RUN_WORKERS not found in environment, using {}."
+                .format(nworker)
+            )
+        if nworker > nproc:
+            log.warning(
+                "Number of workers ({}) larger than number of procs ({})"
+                .format(nworker, nproc)
+            )
+            log.warning(
+                "This should never happen!  Reducing workers to {}"
+                .format(nproc)
+            )
+            nworker = nproc
+        taskproc = nproc // nworker
```

Comment:
Changing this line to 
```python
taskproc = min(nproc // nworker, task_classes[tasktype].run_max_procs())
```
fixes a problem where serial tasks (like fiberflatnight) could be given communicators with multiple ranks that all tried to do the same thing and would clobber each other.  That worked for fixing the fiberflatnight step; I'm now running a larger production from the beginning with that one-line change and will report back if this has some other unintended consequence.

Test production at NERSC in `/global/cscratch1/sd/sjbailey/desi/svdc2019d/spectro/redux/v11`

Reply:
The larger test is still running but so far the only problems have been the logfile redirection timeouts and problems that cascade from there due to dependencies.  I'll commit this change.

Reply:
The logic in run.py around this section is wrong, and I had a local commit Saturday on cori to fix it, but did not push because there was another issue...  I'll try to reconcile this with my changes and open a PR against this branch.